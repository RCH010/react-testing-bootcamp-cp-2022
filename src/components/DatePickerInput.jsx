import React from 'react';

import TextField from '@mui/material/TextField';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { DatePicker } from '@mui/x-date-pickers/DatePicker';

export const DatePickerInput = ({
  value = new Date(),
  setValue = () => {},
  validDate = true,
  setValidDate = () => {},
  maxDate = new Date(),
  minDate = new Date(0),
}) => {
  const handleDateChange = (selectedDate) => {
    let dateNumber = selectedDate && Date.parse(selectedDate);
    const max = Date.parse(maxDate);
    const min = Date.parse(minDate);

    setValidDate(!isNaN(dateNumber) && dateNumber < max && dateNumber > min);
    setValue(selectedDate);
  };

  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <DatePicker
        maxDate={maxDate}
        minDate={minDate}
        label='Pick a date'
        value={value}
        onChange={handleDateChange}
        renderInput={(params) => (
          <TextField
            {...params}
            size='small'
            helperText={!validDate ? 'Please provide a valid date' : ''}
          />
        )}
      />
    </LocalizationProvider>
  );
};
