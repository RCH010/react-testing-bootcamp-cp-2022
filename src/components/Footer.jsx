import { Box, Typography } from '@mui/material';
import React from 'react';

export const Footer = () => {
  return (
    <Box component='footer' sx={{ width: '100%', p: 4 }}>
      <Typography variant='body2' textAlign='center'>
        Project created during Wizeline Academy React Testing Bootcamp
      </Typography>
    </Box>
  );
};
