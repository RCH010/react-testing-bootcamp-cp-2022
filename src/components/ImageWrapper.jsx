import React from 'react';
import PropTypes from 'prop-types';
import { Box, Grid, Skeleton, Typography } from '@mui/material';

const LinesOfSkeletons = () =>
  [...Array(6).keys()].map((index) => <Skeleton variant='text' key={index} />);

export const ImageWrapper = ({
  isLoading,
  title,
  currentDate,
  imageSRC,
  description,
}) => {
  return (
    <Grid
      container
      alignItems='center'
      spacing={4}
      justifyContent='space-between'
      direction={{ xs: 'column', md: 'row' }}>
      <Grid item xs={12} md={6} sx={{ width: '100%' }}>
        <Typography variant='h4' component='h2'>
          {isLoading ? <Skeleton /> : title}
        </Typography>
        <Typography variant='body2'>
          {isLoading ? <Skeleton /> : currentDate}
        </Typography>
        <Box sx={{ '& img': { width: '100%' } }}>
          {isLoading ? (
            <Skeleton variant='rectangular' height='25em' />
          ) : (
            <img src={imageSRC} alt={title} />
          )}
        </Box>
      </Grid>
      <Grid item xs={12} md={6} sx={{ width: '100%' }}>
        <Typography variant='body1'>
          {isLoading ? <LinesOfSkeletons /> : description}
        </Typography>
      </Grid>
    </Grid>
  );
};

ImageWrapper.propTypes = {
  title: PropTypes.string,
  currentDate: PropTypes.string,
  imageSRC: PropTypes.string,
  description: PropTypes.string,
};
