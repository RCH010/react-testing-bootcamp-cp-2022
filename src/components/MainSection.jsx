import React, { useEffect, useState } from 'react';

import { DatePickerInput } from './DatePickerInput';
import { ImageWrapper } from './ImageWrapper';
import { fetchPictureOfTheDay } from '../api/apod';

import { Button, Stack, Typography, Box } from '@mui/material';

export const MainSection = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [apodData, setApodData] = useState(null);
  const [errorMessage, setErrorMessage] = useState(null);
  const [apodDate, setapodDate] = useState(null);
  const [validDate, setValidDate] = useState(new Date());

  useEffect(() => {
    const fetchImage = async () => {
      if (isLoading) {
        try {
          const response = await fetchPictureOfTheDay(apodDate);
          setApodData(response);
          setIsLoading(false);
          setErrorMessage(null);
        } catch (error) {
          setErrorMessage(error.msg || 'There was an error, please try again');
          setIsLoading(false);
        }
      }
    };
    fetchImage();
  }, [isLoading]);

  const onSearchNewDate = (e) => {
    e.preventDefault();
    if (validDate) {
      setIsLoading(true);
    }
  };

  return (
    <Box sx={{ px: 4, flexGrow: 1 }}>
      <Stack
        component='form'
        onSubmit={onSearchNewDate}
        direction='row'
        spacing={4}
        justifyContent='flex-end'
        sx={{ mb: 4 }}>
        <DatePickerInput
          value={apodDate}
          setValue={setapodDate}
          validDate={validDate}
          setValidDate={setValidDate}
        />
        <Button
          variant='contained'
          type='submit'
          disabled={!validDate}
          size='small'>
          Search
        </Button>
      </Stack>
      {errorMessage ? (
        <Box>
          <Typography variant='h3' component='h2'>
            Oh no!
          </Typography>
          <Typography>{errorMessage}</Typography>
        </Box>
      ) : (
        <ImageWrapper
          isLoading={isLoading}
          title={apodData?.title}
          description={apodData?.explanation}
          imageSRC={apodData?.url}
          currentDate={apodData?.date}
        />
      )}
    </Box>
  );
};
