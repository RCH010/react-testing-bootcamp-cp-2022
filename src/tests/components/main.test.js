import {
  fireEvent,
  getByTestId,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MainSection } from '../../components/MainSection';
import testResponse from '../mocks/testResponse.json';

describe('Test for Main section', () => {
  test('should have a date input element', () => {
    render(<MainSection />);

    const dateInputEl = screen.getByLabelText(/pick a date/i);
    expect(dateInputEl).toBeInTheDocument();
  });

  test('should have a submit button', () => {
    render(<MainSection />);

    const submitButton = screen.getByRole('button', { name: /search/i });
    expect(submitButton).toBeInTheDocument();
  });

  test('should have an image on the main section', async () => {
    render(<MainSection />);

    const imageEl = await screen.findByRole('img');
    expect(imageEl).toBeInTheDocument();
  });

  test('should have the image title on the main section', async () => {
    render(<MainSection />);

    const titleEl = await screen.findByRole('heading', {
      name: testResponse.title,
    });
    expect(titleEl).toBeInTheDocument();
  });

  test('should have text for an image description', async () => {
    render(<MainSection />);

    const descriptionEl = await screen.findByText((content, element) =>
      content.startsWith(testResponse.explanation.substring(0, 10))
    );
    expect(descriptionEl).toBeInTheDocument();
  });

  test('should match with snapshot for skeleton purposes', () => {
    const { container } = render(<MainSection />);
    expect(container).toMatchSnapshot();
  });
});
