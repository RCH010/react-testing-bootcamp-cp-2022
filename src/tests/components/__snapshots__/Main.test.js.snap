// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Main section should match with snapshot for skeleton purposes 1`] = `
<div>
  <main
    class="MuiBox-root css-wxwvyg"
  >
    <form
      class="css-jut0lt-MuiStack-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r8:"
          id=":r8:-label"
        >
          Pick a date
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            aria-label="Choose date"
            aria-readonly="true"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r8:"
            readonly=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Pick a date
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root  css-11qr2p8-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Search
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-direction-xs-[object Object] css-7ec0a8-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1qpfcgj-MuiGrid-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </p>
        <div
          class="MuiBox-root css-1pwwlss"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-58ftfq-MuiSkeleton-root"
            style="height: 25em;"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1qpfcgj-MuiGrid-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </p>
      </div>
    </div>
  </main>
</div>
`;
